src/librustc_typeck/check/compare_method.rs

```diff
@@ -138,14 +138,48 @@ pub fn compare_impl_method<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
     }
 
     if impl_m.fty.sig.0.inputs.len() != trait_m.fty.sig.0.inputs.len() {
-        span_err!(tcx.sess, impl_m_span, E0050,
+        let impl_m_node_id = tcx.map.as_local_node_id(impl_m.def_id).unwrap();
+        let mspan = match tcx.map.expect_impl_item(impl_m_node_id).node {
+            ImplItemKind::Method(ref impl_m_sig, _) => {
+                if let Some(arg) = impl_m_sig.decl.inputs.last() {
+                    arg.pat.span
+                } else {
+                    impl_m_span
+                }
+            }
+            _ => bug!("{:?} is not a method", impl_m)
+        };
+        let span = impl_m_span;
+        let (diff, text) = if impl_m.fty.sig.0.inputs.len() > trait_m.fty.sig.0.inputs.len() {
+            (impl_m.fty.sig.0.inputs.len() - trait_m.fty.sig.0.inputs.len(), "less")
+        } else {
+            (trait_m.fty.sig.0.inputs.len() - impl_m.fty.sig.0.inputs.len(), "more")
```

Comment:
nit: you could put `impl_m.fty.sig.0.inputs.len()` and `trait_m.fty.sig.0.inputs.len()` in variables rather than repeating them